src/java/org/apache/cassandra/db/monitoring/MonitoringTask.java

```diff
@@ -328,31 +379,126 @@ protected abstract static class Operation
             numTimesReported = 1;
             totalTimeNanos = failedAtNanos - operation.creationTimeNanos();
             minTime = totalTimeNanos;
             maxTime = totalTimeNanos;
+            timestamp = Clock.Global.currentTimeMillis();
         }
 
+        void add(Operation operation)
+        {
+            numTimesReported++;
+            totalTimeNanos += operation.totalTimeNanos;
+            maxTime = Math.max(maxTime, operation.maxTime);
+            minTime = Math.min(minTime, operation.minTime);
+        }
+
+        public abstract String getLogMessage();
+
+        @JsonProperty
         public String name()
         {
             if (name == null)
                 name = operation.name();
             return name;
         }
 
-        void add(Operation operation)
+        @JsonProperty
+        public String keyspace()
         {
-            numTimesReported++;
-            totalTimeNanos += operation.totalTimeNanos;
-            maxTime = Math.max(maxTime, operation.maxTime);
-            minTime = Math.min(minTime, operation.minTime);
+            if (operation != null)
+            {
+                String monitored = operation.monitoredOnKeyspace();
+                if (monitored != null)
+                    return monitored;
+            }
+            return keyspace;
         }
 
-        public abstract String getLogMessage();
+        @JsonProperty
+        public String table()
+        {
+            if (operation != null)
+            {
+                String monitored = operation.monitoredOnTable();
+                if (monitored != null)
+                    return monitored;
+            }
+            return table;
+        }
+
+        @JsonProperty
+        public boolean isCrossNode()
+        {
+            if (operation != null)
+                return operation.isCrossNode();
+
+            return crossNode;
+        }
+
+        @JsonProperty
+        public int numTimesReported()
+        {
+            return numTimesReported;
+        }
+
+        @JsonProperty
+        public long totalTimeNanos()
+        {
+            return totalTimeNanos;
+        }
+
+        @JsonProperty
+        public long maxTime()
+        {
+            return maxTime;
+        }
+
+        @JsonProperty
+        public long minTime()
+        {
+            return minTime;
+        }
+
+        @JsonIgnore
+        public long averageTime()
+        {
+            return totalTimeNanos / numTimesReported;
+        }
+
+        @JsonProperty
+        public long timestamp()
+        {
+            return timestamp;
+        }
+
+        public static String serialize(Collection<Operation> operations)
+        {
+            return JsonUtils.writeAsJsonString(operations);
+        }
+
+        public static List<Operation> deserialize(String message) throws Throwable
+        {
+            return JsonUtils.JSON_OBJECT_MAPPER.readValue(message, new TypeReference<>() {});
```

Comment:
I suppose TypeReference value can be cached and not allocated for each request..